website/utils/media/processors.py

```diff
@@ -42,22 +42,32 @@ def process_upload(image: DaVinciImage, **kwargs):
 def thumbnail(image: DaVinciImage, **kwargs):
     """Thumbnail an image to at most the given size.
 
-    If `kwargs["cover"]`, the image is cropped to the aspect ratio of the given size.
+    If `mode="contain"` (the default), the image is resized to fit within the given size.
+    The original aspect ratio is preserved, so the image may be smaller than specified.
+
+    If `mode="cover"`, the image is cropped to the aspect ratio of the given size.
     The thumbnails are saved to WebP format with lossy compression.
 
+    If `mode="pad"`, the image is resized to fit within the given size.
+    Padding is added to the edges to make the image the exact size specified.
+
     Warning: for django-thumbnails to save the image with the right filename,
     the '<size>.FORMAT' key in settings.THUMBNAILS_SIZES must be set to 'webp'.
 
     """
     pil_image = image.get_pil_image()
 
     size = kwargs["size"]
-    if not kwargs.get("cover", False):
-        ratio = min(a / b for a, b in zip(size, pil_image.size))
-        size = tuple(int(ratio * x) for x in pil_image.size)
-
-    if size[0] != pil_image.size[0] and size[1] != pil_image.size[1]:
-        pil_image = ImageOps.fit(pil_image, size, Image.Resampling.LANCZOS)
+    mode = kwargs.get("mode", "contain")
+
+    match mode:
+        case "contain":
+            pil_image = ImageOps.contain(pil_image, size, Image.Resampling.LANCZOS)
+        case "cover":
+            pil_image = ImageOps.fit(pil_image, size, Image.Resampling.LANCZOS)
```

Comment:
This changes the behaviour of the cover mode a bit, at least by scaling up if the original is smaller, but maybe also in other ways. So I'd prefer to keep that case unchanged. 
But love the match case stuff!

Reply:
What do you mean? I think that `cover` just calls `ImageOps.fit`, same as this PR.

For `contain` aka `cover=False` you are right that `ImageOps.contain` replaces our custom logic. But I don't see how our old version prevents scaling up; won't that just give a ratio greater than 1 and also scale up?